### ISSUE TITLE
Low loss, high recall(pjreddie) but 0 mAP(AlexeyAB) 

### ISSUE THREAD

**Issue by CKsunvictor7** [AlexeyAB/darknet#1179]
Hi AlexeyAB and all, 

I trained yolov3 on my customized DB(256 categories) using pjreddie's version.  
after long training, I got this curve 
![uec256](https://user-images.githubusercontent.com/14801061/42562652-40db82ba-8537-11e8-86ec-0e4993aabc62.png). 

then I evaluated using 'recall'(pjreddie) on (final.weights):
    0     1     1	RPs/Img: 10.00	IOU: 55.33%	Recall:100.00%
    1     2     2	RPs/Img: 10.50	IOU: 59.13%	Recall:100.00%
    2     3     3	RPs/Img: 12.00	IOU: 58.19%	Recall:100.00%
    3     3     4	RPs/Img: 13.00	IOU: 43.64%	Recall:75.00%
    4     4     5	RPs/Img: 13.60	IOU: 47.83%	Recall:80.00%
    5     5     6	RPs/Img: 14.50	IOU: 51.26%	Recall:83.33%
    6     5     7	RPs/Img: 15.57	IOU: 50.77%	Recall:71.43%
    7     6     8	RPs/Img: 15.25	IOU: 52.90%	Recall:75.00%
    8     7     9	RPs/Img: 16.56	IOU: 52.91%	Recall:77.78%
    9     8    10	RPs/Img: 16.50	IOU: 54.06%	Recall:80.00%
   10     8    11	RPs/Img: 17.18	IOU: 53.62%	Recall:72.73%
... 
somehow I got 'Segmentation fault (core dumped)' at 121 categories.
->
So I did evaluation using AlexeyAB's 'map' on (final.weights, 900.weights) and both got ... 
...
class_id = 245, name = eggplant with garlic sauce, 	 ap = 0.00 % 
class_id = 246, name = three cup chicken, 	 ap = 0.00 % 
class_id = 247, name = bean curd family style, 	 ap = 0.00 % 
class_id = 248, name = salt & pepper fried shrimp with shell, 	 ap = 0.00 % 
class_id = 249, name = baked salmon, 	 ap = 0.00 % 
class_id = 250, name = braised pork meat ball with napa cabbage, 	 ap = 0.00 % 
class_id = 251, name = winter melon soup, 	 ap = 0.00 % 
class_id = 252, name = steamed spareribs, 	 ap = 0.00 % 
class_id = 253, name = chinese pumpkin pie, 	 ap = 0.00 % 
class_id = 254, name = eight treasure rice, 	 ap = 0.00 % 
class_id = 255, name = hot & sour soup, 	 ap = 0.00 % 
 for thresh = 0.25, precision = -nan, recall = 0.00, F1-score = -nan 
 for thresh = 0.25, TP = 0, FP = 0, FN = 3156, average IoU = 0.00 % 
 mean average precision (mAP) = 0.000071, or 0.01 % .... 

And detection results are same as mAP(AlexeyAB), means detect nothing for all data (even on training data).

------
Wondering how to solve those problems:
1. what is the reason caused the difference between recall(pjreddie) and mAP(AlexeyAB).

2. what are the reasons may cause low loss but almost 0 mAP... 

3. The size of data varies, from 120x180 to 640x480.  Is it ok to set width and height as 480, 480?  

------
What I have done:
1. recalculate anchors on all dataset: got 140.4457,155.3769, 206.4959,243.0188, 374.2764,228.2118, 278.9701,337.0385, 373.2221,353.6263, 453.0368,324.6729, 363.6068,432.5764, 438.5675,411.1638, 459.8292,458.4581. 

2. checked labels and I am sure from 0~255, each classes has data

3. my data was collected by another labeling tool. But the YOLO format is right(x, y, w, h). And ok for previous  training test.

4. changed the value of classes and filters in .cfg 

Thank you all!

**Comment by AlexeyAB** [comment]
@CKsunvictor7 Hi,

> what is the reason caused the difference between recall(pjreddie) and mAP(AlexeyAB).

* recall(pjreddie) - are internal indicators for developers only

* mAP(AlexeyAB) - are the most used indicators in public competitions, can be used for users and for comparison with other models

> what are the reasons may cause low loss but almost 0 mAP...

Training goes wrong

> The size of data varies, from 120x180 to 640x480. Is it ok to set width and height as 480, 480?

it is ok.

----

* Check your dataset by using Yolo_mark: https://github.com/AlexeyAB/Yolo_mark

* Show your cfg-file.

* Try to train ~5000-10000 iterations by using this repository, and then check mAP.

**Comment by CKsunvictor7** [comment]
1. seems it looks ok by using Yolo_mark
![image](https://user-images.githubusercontent.com/14801061/42566460-a4f02f08-8541-11e8-9f33-80e29c42bac7.png)

2. my cfg:
[net]
# Testing
 batch=1
 subdivisions=1
# Training
batch=64
subdivisions=4
width=480
height=480
channels=3
momentum=0.9
decay=0.0005
angle=90
saturation = .0
exposure = .0
hue=.0

learning_rate=0.00001
burn_in=1000
max_batches = 150000
policy=steps
steps=60000,90000,120000,130000
scales=.5,.5,.5,.1

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

# Downsample

[convolutional]
batch_normalize=1
filters=64
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=32
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=128
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=256
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=512
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

######################

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=783
activation=linear

[yolo]
mask = 6,7,8
anchors = 140.4457,155.3769, 206.4959,243.0188, 374.2764,228.2118, 278.9701,337.0385, 373.2221,353.6263, 453.0368,324.6729, 363.6068,432.5764, 438.5675,411.1638, 459.8292,458.4581
classes=256
num=9
jitter=.3
ignore_thresh = .5
truth_thresh = 1
random=1

[route]
layers = -4

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 61



[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=512
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=512
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=512
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=783
activation=linear

[yolo]
mask = 3,4,5
anchors = 140.4457,155.3769, 206.4959,243.0188, 374.2764,228.2118, 278.9701,337.0385, 373.2221,353.6263, 453.0368,324.6729, 363.6068,432.5764, 438.5675,411.1638, 459.8292,458.4581
classes=256
num=9
jitter=.3
ignore_thresh = .5
truth_thresh = 1
random=1

[route]
layers = -4

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 36



[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=783
activation=linear

[yolo]
mask = 0,1,2
anchors = 140.4457,155.3769, 206.4959,243.0188, 374.2764,228.2118, 278.9701,337.0385, 373.2221,353.6263, 453.0368,324.6729, 363.6068,432.5764, 438.5675,411.1638, 459.8292,458.4581
classes=256
num=9
jitter=.3
ignore_thresh = .5
truth_thresh = 1
random=1
 
3. OK.  I am going to do it.  Thank you very much!!

**Comment by AlexeyAB** [comment]
1. Comment or remove lines:
```
 batch=1
 subdivisions=1
```

2. If you want to disable color data augmentation, then set
```
saturation = 1.0
exposure = 1.0
hue=.0
```

But better to enable color data augmentation, at least exposure:
```
saturation = 1.0
exposure = 1.5
hue=.0
```

3. Try to set higher laerning rate:
```
learning_rate=0.001
```

4. As I see from your anchors, there are no small objects, so try to set
```
width=320
height=320
```

5. Try to train by using this repository.

**Comment by CKsunvictor7** [comment]
Thank you!!  
But I got the messages, when tried to train the models, 
---
./darknet detector train exps/UEC256.data exps/UEC256.cfg darknet53.conv.74 2>1 | tee training_log_UEC.txt
UEC256
Total BFLOPS 89.392 

 seen 64 
Learning Rate: 0.001, Momentum: 0.9, Decay: 0.0005
 If error occurs - run training with flag: -dont_show 
---
and it stopped without message.  
I guess it read no data?  
Should I put images in the same location which you mentioned in Readme?

**Comment by AlexeyAB** [comment]
@CKsunvictor7 


Try to set in cfg-file
```
batch=64
subdivisions=64
```

**Comment by CKsunvictor7** [comment]
@AlexeyAB 
Sorry to bother you again. 

I got the msg and stopped without error msg: 
finc@NC0006:/mnt2/dc_projects/AI_DC_YOLO_AB$ ./darknet detector train exps/UEC256.data exps/UEC256.cfg darknet53.conv.74 2>1 | tee training_log_UEC.txt
UEC256
Total BFLOPS 89.392 

 seen 64 
Learning Rate: 0.001, Momentum: 0.9, Decay: 0.0005
 If error occurs - run training with flag: -dont_show 
==>
Then I ran it with -dont_show, it works.   
./darknet detector train -dont_show  exps/UEC256.data exps/UEC256.cfg darknet53.conv.74 2>1 | tee training_log_UEC.txt 
-------------------------
wondering what -dont_show is?

**Comment by AlexeyAB** [comment]
@CKsunvictor7 By default my repository shows Avg Loss chart, but if there is no monitor, then it lead to crash. 
`-dont_show` means that Avg Loss chart will not be showed.
